<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="vcfppR">
<title>Assess the concordance between VCFs • vcfppR</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Assess the concordance between VCFs">
<meta property="og:description" content="vcfppR">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">vcfppR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/concordance-of-two-vcf-files.html">Assess the concordance between VCFs</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Zilong-Li/vcfppR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="concordance-of-two-vcf-files_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Assess the concordance between VCFs</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/Zilong-Li/vcfppR/blob/HEAD/vignettes/concordance-of-two-vcf-files.Rmd" class="external-link"><code>vignettes/concordance-of-two-vcf-files.Rmd</code></a></small>
      <div class="d-none name"><code>concordance-of-two-vcf-files.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/Zilong-Li/vcfppR" class="external-link">vcfppR</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h2>
<p>In a benchmarking, it’s often that we need to calculate the concordance rate between the test set and truth set. In the truth VCF file, there are always true genotypes <strong>GT</strong> otherwise we can’t validate the test VCF. A test VCF may be from the variant caller or the genotype imputation program, and the format can be variable, e.g <strong>GP</strong>, <strong>DS</strong> and <strong>GT</strong>. Hence, this article guides you how to use the <code><a href="../reference/vcfcomp.html">vcfppR::vcfcomp</a></code> function to rapidly examine various statistics for different scenarios and formats, such as the Pearson correlation of genotyping (stats=“r2”), the Non-Reference Concordance (stats=“nrc”), the F1-score (stats=“f1”) or the Phasing Switch Error (stats=“pse”).</p>
</div>
<div class="section level2">
<h2 id="case-1-genotype-imputation-accuracy">Case 1: Genotype imputation accuracy<a class="anchor" aria-label="anchor" href="#case-1-genotype-imputation-accuracy"></a>
</h2>
<p>We normally get genotype posterior <code>GP</code> and genotype dosage <code>DS</code> from the diploid imputation software, eg QUILT and GLIMPSE. To examine the imputation accuracy, we calculate the Pearson correlation between the imputed genotype dosage and the true genotypes. With <code>vcfcomp</code>, we need to specify the desired <code>stats="r2"</code> and <code>formats=c("DS","GT")</code>, which will extract the respective FORMAT items for the <code>testvcf</code> and <code>truthvcf</code>.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span><span class="va">testvcf</span>, <span class="va">truthvcf</span>, formats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"DS"</span>, <span class="st">"GT"</span><span class="op">)</span>, stats <span class="op">=</span> <span class="st">"r2"</span><span class="op">)</span></span></code></pre></div>
<p>Besides, the QUILT2-nipt method outputs <code>MDS</code> and <code>FDS</code> for both maternal and fetal genotype dosages in constract to the <code>DS</code> in diploid mode. To assess the imputation accuracy of both the maternal and fetal, we only need to specify the corresponding <code>formats</code>.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span><span class="va">testvcf</span>, <span class="va">truthvcf</span>, formats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MDS"</span>, <span class="st">"GT"</span><span class="op">)</span>, stats <span class="op">=</span> <span class="st">"r2"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span><span class="va">testvcf</span>, <span class="va">truthvcf</span>, formats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FDS"</span>, <span class="st">"GT"</span><span class="op">)</span>, stats <span class="op">=</span> <span class="st">"r2"</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> In the absence of <code>DS</code> format in the testvcf, one can either fill the <code>DS</code> tag using the <a href="https://github.com/Zilong-Li/vcfpp/blob/main/tools" class="external-link">vcf_addDS</a> command line tool or manipulate the testvcf in R using <code>vcftable</code>.</p>
</div>
<div class="section level2">
<h2 id="case-2-genotype-concordance">Case 2: Genotype concordance<a class="anchor" aria-label="anchor" href="#case-2-genotype-concordance"></a>
</h2>
<p>In this case, we are interested in the called genotype concordance and the sensitivity / specificity in genotype calling. In contrast to <code>stats="r2"</code>, we choose <code>stats="f1"</code> or <code>stats="nrc"</code> and specify the <code>formats=c("GT", "GT")</code>. Normally, we want the results for each sample, which can be achieved by using <code>by.sample=TRUE</code>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span><span class="va">testvcf</span>, <span class="va">truthvcf</span>, formats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GT"</span>,<span class="st">"GT"</span><span class="op">)</span>, stats<span class="op">=</span><span class="st">"nrc"</span>, by.sample<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span><span class="va">testvcf</span>, <span class="va">truthvcf</span>, formats <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"GT"</span>,<span class="st">"GT"</span><span class="op">)</span>, stats<span class="op">=</span><span class="st">"f1"</span>, by.sample<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="case-3-phasing-switch-error">Case 3: Phasing switch error<a class="anchor" aria-label="anchor" href="#case-3-phasing-switch-error"></a>
</h2>
<p>In this case, we look for functionality to assess the phasing switch error. First of all, we need the two VCF files to contain the <strong>phased GT</strong>, which is represented through the ‘|’. We can choose to return the sites that have pse.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span><span class="va">testvcf</span>, <span class="va">truthvcf</span>, stats<span class="op">=</span><span class="st">"pse"</span>, return_pse_sites<span class="op">=</span><span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p><strong>Note:</strong> Currently, the <code>pse</code> statatics is a simple form that doesn’t take the completeness and quality into account.</p>
</div>
<div class="section level2">
<h2 id="case-4-multiple-testing-repetitively">Case 4: Multiple testing repetitively<a class="anchor" aria-label="anchor" href="#case-4-multiple-testing-repetitively"></a>
</h2>
<p>In the comprehensive benchmarking, we often run many tests against the same true sets. In this scenario, we can save the truth as an RDS object and reuse it. Actually, both <code>test</code> and <code>truth</code> can take an RDS file as input. The RDS file stores an object that returned by <code>vcftable</code>.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/readRDS.html" class="external-link">saveRDS</a></span><span class="op">(</span><span class="fu"><a href="../reference/vcftable.html">vcftable</a></span><span class="op">(</span><span class="va">truthvcf</span><span class="op">)</span>, <span class="st">"truth.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span>test<span class="op">=</span><span class="va">testvcf1</span>, truth<span class="op">=</span><span class="st">"truth.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span>test<span class="op">=</span><span class="va">testvcf2</span>, truth<span class="op">=</span><span class="st">"truth.rds"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/vcfcomp.html">vcfcomp</a></span><span class="op">(</span>test<span class="op">=</span><span class="va">testvcf3</span>, truth<span class="op">=</span><span class="st">"truth.rds"</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="qa-how-to-fix-the-error-inconsistent-samples-name">Q&amp;A: How to fix the error “inconsistent samples name”<a class="anchor" aria-label="anchor" href="#qa-how-to-fix-the-error-inconsistent-samples-name"></a>
</h2>
<p>If one is certain about the samples name in the testvcf can be replaced with other names that can match the samples in the truthvcf. One can use the <code>names</code> option to specify a vector of new names that can be found in the truthvcf.</p>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Zilong Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
