<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The swiss army knife for reading VCF/BCF into R data types rapidly and easily."><title>read VCF/BCF contents into R data structure — vcftable • vcfppR</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.11/clipboard.min.js" integrity="sha512-7O5pXpc0oCRrxk8RUfDYFgn0nO1t+jLuIOQdOMRp4APB7uZ4vSjspzp5y6YDtDs4VzUSTbWzBFZ/LKJhnyFOKw==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="read VCF/BCF contents into R data structure — vcftable"><meta property="og:description" content="The swiss army knife for reading VCF/BCF into R data types rapidly and easily."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-dark navbar-expand-lg bg-primary" data-bs-theme="dark"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">vcfppR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.4.6</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/concordance-of-two-vcf-files.html">Assess the concordance between VCFs</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/Zilong-Li/vcfppR/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>read VCF/BCF contents into R data structure</h1>
      <small class="dont-index">Source: <a href="https://github.com/Zilong-Li/vcfppR/blob/HEAD/R/vcf-tables.R" class="external-link"><code>R/vcf-tables.R</code></a></small>
      <div class="d-none name"><code>vcftable.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The swiss army knife for reading VCF/BCF into R data types rapidly and easily.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">vcftable</span><span class="op">(</span></span>
<span>  <span class="va">vcffile</span>,</span>
<span>  region <span class="op">=</span> <span class="st">""</span>,</span>
<span>  samples <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>  vartype <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  format <span class="op">=</span> <span class="st">"GT"</span>,</span>
<span>  ids <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  qual <span class="op">=</span> <span class="fl">0</span>,</span>
<span>  pass <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  info <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  collapse <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  setid <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  mac <span class="op">=</span> <span class="fl">0</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>vcffile</dt>
<dd><p>path to the VCF/BCF file</p></dd>


<dt>region</dt>
<dd><p>region to subset in bcftools-like style: "chr1", "chr1:1-10000000"</p></dd>


<dt>samples</dt>
<dd><p>samples to subset in bcftools-like style.
comma separated list of samples to include (or exclude with "^" prefix).
e.g. "id01,id02", "^id01,id02".</p></dd>


<dt>vartype</dt>
<dd><p>restrict to specific type of variants. supports "snps","indels", "sv", "multisnps","multiallelics"</p></dd>


<dt>format</dt>
<dd><p>the FORMAT tag to extract. default "GT" is extracted.</p></dd>


<dt>ids</dt>
<dd><p>character vector. restrict to sites with ID in the given vector. default NULL won't filter any sites.</p></dd>


<dt>qual</dt>
<dd><p>numeric. restrict to variants with QUAL &gt; qual.</p></dd>


<dt>pass</dt>
<dd><p>logical. restrict to variants with FILTER = "PASS".</p></dd>


<dt>info</dt>
<dd><p>logical. drop INFO column in the returned list.</p></dd>


<dt>collapse</dt>
<dd><p>logical. It acts on the FORMAT. If the FORMAT to extract is "GT", the dim of raw genotypes matrix of diploid is (M, 2 * N),
where M is #markers and N is #samples. default TRUE will collapse the genotypes for each sample such that the matrix is (M, N).
Set this to FALSE if one wants to maintain the phasing order, e.g. "1|0" is parsed as c(1, 0) with collapse=FALSE.
If the FORMAT to extract is not "GT", then with collapse=TRUE it will try to turn a list of the extracted vector into a matrix.
However, this raises issues when one variant is mutliallelic resulting in more vaules than others.</p></dd>


<dt>setid</dt>
<dd><p>logical. reset ID column as CHR_POS_REF_ALT.</p></dd>


<dt>mac</dt>
<dd><p>integer. restrict to variants with minor allele count higher than the value.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>Return a list containing the following components:</p><dl><dt>samples</dt>
<dd><p>: character vector; <br>
the samples ids in the VCF file after subsetting</p></dd>


<dt>chr</dt>
<dd><p>: character vector; <br>
the CHR column in the VCF file</p></dd>


<dt>pos</dt>
<dd><p>: character vector; <br>
the POS column in the VCF file</p></dd>


<dt>id</dt>
<dd><p>: character vector; <br>
the ID column in the VCF file</p></dd>


<dt>ref</dt>
<dd><p>: character vector; <br>
the REF column in the VCF file</p></dd>


<dt>alt</dt>
<dd><p>: character vector; <br>
the ALT column in the VCF file</p></dd>


<dt>qual</dt>
<dd><p>: character vector; <br>
the QUAL column in the VCF file</p></dd>


<dt>filter</dt>
<dd><p>: character vector; <br>
the FILTER column in the VCF file</p></dd>


<dt>info</dt>
<dd><p>: character vector; <br>
the INFO column in the VCF file</p></dd>


<dt>format</dt>
<dd><p>: matrix of either integer of numberic values depending on the tag to extract; <br>
a specifiy tag in the FORMAT column to be extracted</p></dd>


</dl></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p><code>vcftable</code> uses the C++ API of vcfpp, which is a wrapper of htslib, to read VCF/BCF files.
Thus, it has the full functionalities of htslib, such as restrict to specific variant types,
samples and regions. For the memory efficiency reason, the <code>vcftable</code> is designed
to parse only one tag at a time in the FORMAT column of the VCF. In default, only the matrix of genotypes,
i.e. "GT" tag, are returned by <code>vcftable</code>, but there are many other tags supported by the <code>format</code> option.</p>
    </div>
    <div class="section level2">
    <h2 id="author">Author<a class="anchor" aria-label="anchor" href="#author"></a></h2>
    <p>Zilong Li <a href="mailto:zilong.dk@gmail.com">zilong.dk@gmail.com</a></p>
    </div>

    <div class="section level2">
    <h2 id="ref-examples">Examples<a class="anchor" aria-label="anchor" href="#ref-examples"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://github.com/Zilong-Li/vcfppR" class="external-link">'vcfppR'</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vcffile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"raw.gt.vcf.gz"</span>, package<span class="op">=</span><span class="st">"vcfppR"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">vcftable</span><span class="op">(</span><span class="va">vcffile</span>, <span class="st">"chr21:1-5050000"</span>, vartype <span class="op">=</span> <span class="st">"snps"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> List of 10</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ samples: chr [1:3202] "HG00096" "HG00097" "HG00099" "HG00100" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ chr    : chr [1:12] "chr21" "chr21" "chr21" "chr21" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ pos    : int [1:12] 5030082 5030088 5030105 5030253 5030278 5030347 5030356 5030357 5030391 5030446 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ id     : chr [1:12] "chr21:5030082:G:A" "chr21:5030088:C:T" "chr21:5030105:C:A" "chr21:5030253:G:T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ ref    : chr [1:12] "G" "C" "C" "G" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ alt    : chr [1:12] "A" "T" "A" "T" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ qual   : num [1:12] 70.1 2773.1 3897.8 102.6 868.9 ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ filter : chr [1:12] "VQSRTrancheSNP99.80to100.00" "VQSRTrancheSNP99.80to100.00" "VQSRTrancheSNP99.80to100.00" "VQSRTrancheSNP99.80to100.00" ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ info   : chr [1:12] "AC=2;AF=0.000616523;AN=3244;DP=2498;FS=0;MLEAC=1;MLEAF=0.0003083;MQ=17.07;MQ0=0;QD=17.52;SOR=3.258;VQSLOD=-32.6"| __truncated__ "AC=127;AF=0.0400126;AN=3174;BaseQRankSum=0.736;ClippingRankSum=0.736;DP=2750;FS=0;InbreedingCoeff=0.0015;MLEAC="| __truncated__ "AC=128;AF=0.0352811;AN=3628;BaseQRankSum=0.736;ClippingRankSum=0.727;DP=3476;FS=0;InbreedingCoeff=-0.0015;MLEAC"| __truncated__ "AC=1;AF=0.000165837;AN=6030;BaseQRankSum=-0.583;ClippingRankSum=-0.259;DP=19530;FS=0;InbreedingCoeff=-0.0274;ML"| __truncated__ ...</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>  $ gt     : int [1:12, 1:3202] NA 2 2 0 0 0 0 0 0 0 ...</span>
</code></pre></div>
    </div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p>Developed by Zilong Li.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

    </footer></div>

  

  

  </body></html>

