# Generated by using Rcpp::compileAttributes() -> do not edit by hand
# Generator token: 10BE3573-1514-4C36-9D1C-5A225CD40393

#' get DS (dosage) of each sample and skip variants given a INFO tag of Float type and a cutoff
#' @param cutoff  skip variants with tag value smaller than this cutoff
#' @param tag     tag exists in INFO column
#' @param vcffile path to the VCF file with index
#' @param region  region to extract
#' @param samples samples to extract
#' @return A list of genotype dosages for each sample along with variant information in VCF
#' @export
getDSskipInfoTagFloat <- function(cutoff, tag, vcffile, region = "", samples = "-") {
    .Call(`_vcfppR_getDSskipInfoTagFloat`, cutoff, tag, vcffile, region, samples)
}

#' calculate the number of heterozygous SNPs for each sample
#' @param vcffile path to the VCF file with index
#' @param region  region to extract, default "" for all
#' @param pass restrict to variants with FILTER==PASS
#' @param qual restrict to variants with QUAL > qual.
#' @param samples samples to extract, default "-" for all
#' @return A list of heterozygosity couts for each sample along with its id in the vcf header
#' @export
heterozygosity <- function(vcffile, region = "", samples = "-", pass = FALSE, qual = 0) {
    .Call(`_vcfppR_heterozygosity`, vcffile, region, samples, pass, qual)
}

tableNA <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableNA`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableGT <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableGT`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableGP <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableGP`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableDS <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableDS`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableGL <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableGL`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableAD <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableAD`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tablePL <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tablePL`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableGQ <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableGQ`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableHQ <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableHQ`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableDP <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableDP`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tableMQ <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tableMQ`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

tablePQ <- function(vcffile, region, samples = "-", qualval = 0, pass = FALSE, INFO = TRUE, snps = FALSE, indels = FALSE, multiallelics = FALSE, multisnps = FALSE) {
    .Call(`_vcfppR_tablePQ`, vcffile, region, samples, qualval, pass, INFO, snps, indels, multiallelics, multisnps)
}

summaryVariants <- function(vcffile, region = "", samples = "-", filter_pass = FALSE, qual = 0) {
    .Call(`_vcfppR_summaryVariants`, vcffile, region, samples, filter_pass, qual)
}

summarySVs <- function(vcffile, region = "", samples = "-", filter_pass = FALSE, qual = 0) {
    .Call(`_vcfppR_summarySVs`, vcffile, region, samples, filter_pass, qual)
}

